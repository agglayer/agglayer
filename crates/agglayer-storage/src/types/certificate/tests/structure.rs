use agglayer_types::{Address, Digest, NetworkId};
use pessimistic_proof::unified_bridge;

fn trace() -> serde_reflection::Result<serde_reflection::Registry> {
    let tracer_config = serde_reflection::TracerConfig::default().is_human_readable(false);
    let mut tracer = serde_reflection::Tracer::new(tracer_config);
    let mut samples = serde_reflection::Samples::new();

    // Support types with custom values.
    dbg!(tracer.trace_value::<Address>(&mut samples, &Address::new([0xad; 20]))?);
    dbg!(tracer.trace_value::<Digest>(&mut samples, &Digest([0xd1; 32]))?);
    dbg!(tracer.trace_value::<NetworkId>(&mut samples, &NetworkId::new(42))?);

    // Remaining support types.
    dbg!(tracer.trace_type::<agglayer_types::GenerationType>(&samples)?);
    dbg!(tracer.trace_type::<unified_bridge::GlobalIndex>(&samples)?);
    dbg!(tracer.trace_type::<unified_bridge::TokenInfo>(&samples)?);

    // Error enumerations.
    dbg!(tracer.trace_type::<unified_bridge::LocalExitTreeError>(&samples)?);
    dbg!(tracer.trace_type::<unified_bridge::Error>(&samples)?);
    dbg!(tracer.trace_type::<pessimistic_proof::ProofError>(&samples)?);
    dbg!(tracer.trace_type::<pessimistic_proof::error::ProofVerificationError>(&samples)?);
    dbg!(tracer.trace_type::<agglayer_tries::error::SmtError>(&samples)?);
    dbg!(tracer.trace_type::<agglayer_types::Error>(&samples)?);
    dbg!(tracer.trace_type::<agglayer_types::CertificateStatusError>(&samples)?);

    // Certificate status.
    dbg!(tracer.trace_type::<agglayer_types::CertificateStatus>(&samples)?);

    tracer.registry()
}

#[test]
fn structure_snapshot() {
    // Take a snapshot of the internal structure of the encoding of
    // Certificate-related types to avoid accidental changes.
    //
    // Note that if this changes, backwards compatibility of the change has to
    // be considered. As a rule of thumb:
    // * It is OK to add an enum arm at the end
    // * It is NOT OK to remove, reorder or change existing enum arms.
    // * It is NOT OK to alter structs in any way, even just by adding a field
    //   at the end.
    match trace() {
        Ok(registry) => insta::assert_json_snapshot!(&registry),
        Err(err) => panic!("{err}: {}", err.explanation()),
    }
}
