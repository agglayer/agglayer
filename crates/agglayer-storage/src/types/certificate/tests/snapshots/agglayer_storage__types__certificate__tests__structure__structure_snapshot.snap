---
source: crates/agglayer-storage/src/types/certificate/tests/structure.rs
expression: "&registry"
snapshot_kind: text
---
{
  "CertificateStatus": {
    "ENUM": {
      "0": {
        "Pending": "UNIT"
      },
      "1": {
        "Proven": "UNIT"
      },
      "2": {
        "Candidate": "UNIT"
      },
      "3": {
        "InError": {
          "STRUCT": [
            {
              "error": {
                "TYPENAME": "CertificateStatusError"
              }
            }
          ]
        }
      },
      "4": {
        "Settled": "UNIT"
      }
    }
  },
  "CertificateStatusError": {
    "ENUM": {
      "0": {
        "ProofGenerationError": {
          "STRUCT": [
            {
              "generation_type": {
                "TYPENAME": "GenerationType"
              }
            },
            {
              "source": {
                "TYPENAME": "ProofError"
              }
            }
          ]
        }
      },
      "1": {
        "ProofVerificationFailed": {
          "NEWTYPE": {
            "TYPENAME": "ProofVerificationError"
          }
        }
      },
      "2": {
        "TypeConversionError": {
          "NEWTYPE": {
            "TYPENAME": "agglayer_types::Error"
          }
        }
      },
      "3": {
        "TrustedSequencerNotFound": {
          "NEWTYPE": {
            "TYPENAME": "NetworkId"
          }
        }
      },
      "4": {
        "InternalError": {
          "NEWTYPE": "STR"
        }
      },
      "5": {
        "SettlementError": {
          "NEWTYPE": "STR"
        }
      },
      "6": {
        "PreCertificationError": {
          "NEWTYPE": "STR"
        }
      },
      "7": {
        "CertificationError": {
          "NEWTYPE": "STR"
        }
      },
      "8": {
        "L1InfoRootNotFound": {
          "NEWTYPE": "U32"
        }
      },
      "9": {
        "LastPessimisticRootNotFound": {
          "NEWTYPE": {
            "TYPENAME": "NetworkId"
          }
        }
      }
    }
  },
  "GenerationType": {
    "ENUM": {
      "0": {
        "Native": "UNIT"
      },
      "1": {
        "Prover": "UNIT"
      }
    }
  },
  "GlobalIndex": {
    "STRUCT": [
      {
        "mainnet_flag": "BOOL"
      },
      {
        "rollup_index": "U32"
      },
      {
        "leaf_index": "U32"
      }
    ]
  },
  "LocalExitTreeError": {
    "ENUM": {
      "0": {
        "LeafIndexOverflow": "UNIT"
      },
      "1": {
        "IndexOutOfBounds": "UNIT"
      },
      "2": {
        "FrontierIndexOutOfBounds": "UNIT"
      }
    }
  },
  "NetworkId": {
    "NEWTYPESTRUCT": "U32"
  },
  "ProofError": {
    "ENUM": {
      "0": {
        "InvalidPreviousLocalExitRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "1": {
        "InvalidPreviousBalanceRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "2": {
        "InvalidPreviousNullifierRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "3": {
        "InvalidNewLocalExitRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "4": {
        "InvalidNewBalanceRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "5": {
        "InvalidNewNullifierRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "6": {
        "InvalidImportedBridgeExit": {
          "STRUCT": [
            {
              "source": {
                "TYPENAME": "unified_bridge::Error"
              }
            },
            {
              "global_index": {
                "TYPENAME": "GlobalIndex"
              }
            }
          ]
        }
      },
      "7": {
        "InvalidImportedExitsRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "8": {
        "MismatchImportedExitsRoot": "UNIT"
      },
      "9": {
        "InvalidNullifierPath": "UNIT"
      },
      "10": {
        "InvalidBalancePath": "UNIT"
      },
      "11": {
        "BalanceOverflowInBridgeExit": "UNIT"
      },
      "12": {
        "BalanceUnderflowInBridgeExit": "UNIT"
      },
      "13": {
        "CannotExitToSameNetwork": "UNIT"
      },
      "14": {
        "InvalidMessageOriginNetwork": "UNIT"
      },
      "15": {
        "InvalidL1TokenInfo": {
          "NEWTYPE": {
            "TYPENAME": "TokenInfo"
          }
        }
      },
      "16": {
        "MissingTokenBalanceProof": {
          "NEWTYPE": {
            "TYPENAME": "TokenInfo"
          }
        }
      },
      "17": {
        "DuplicateTokenBalanceProof": {
          "NEWTYPE": {
            "TYPENAME": "TokenInfo"
          }
        }
      },
      "18": {
        "InvalidSignature": "UNIT"
      },
      "19": {
        "InvalidSigner": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Address"
              }
            },
            {
              "recovered": {
                "TYPENAME": "agglayer_primitives::Address"
              }
            }
          ]
        }
      },
      "20": {
        "InvalidLocalExitTreeOperation": {
          "NEWTYPE": {
            "TYPENAME": "LocalExitTreeError"
          }
        }
      },
      "21": {
        "Unknown": {
          "NEWTYPE": "STR"
        }
      },
      "22": {
        "InvalidPreviousPessimisticRoot": {
          "STRUCT": [
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed_v2": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "computed_v3": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "23": {
        "InconsistentSignedPayload": "UNIT"
      },
      "24": {
        "HeightOverflow": "UNIT"
      }
    }
  },
  "ProofVerificationError": {
    "ENUM": {
      "0": {
        "VersionMismatch": {
          "NEWTYPE": "STR"
        }
      },
      "1": {
        "Core": {
          "NEWTYPE": "STR"
        }
      },
      "2": {
        "Recursion": {
          "NEWTYPE": "STR"
        }
      },
      "3": {
        "Plonk": {
          "NEWTYPE": "STR"
        }
      },
      "4": {
        "Groth16": {
          "NEWTYPE": "STR"
        }
      },
      "5": {
        "InvalidPublicValues": "UNIT"
      },
      "6": {
        "Other": {
          "NEWTYPE": "STR"
        }
      }
    }
  },
  "SmtError": {
    "ENUM": {
      "0": {
        "KeyAlreadyPresent": "UNIT"
      },
      "1": {
        "KeyNotPresent": "UNIT"
      },
      "2": {
        "KeyPresent": "UNIT"
      },
      "3": {
        "DepthOutOfBounds": "UNIT"
      }
    }
  },
  "TokenInfo": {
    "STRUCT": [
      {
        "origin_network": {
          "TYPENAME": "NetworkId"
        }
      },
      {
        "origin_token_address": {
          "TYPENAME": "agglayer_primitives::Address"
        }
      }
    ]
  },
  "agglayer_primitives::Address": {
    "NEWTYPESTRUCT": "BYTES"
  },
  "agglayer_primitives::Digest": {
    "NEWTYPESTRUCT": {
      "TUPLEARRAY": {
        "CONTENT": "U8",
        "SIZE": 32
      }
    }
  },
  "agglayer_types::Error": {
    "ENUM": {
      "0": {
        "MultipleL1InfoRoot": "UNIT"
      },
      "1": {
        "MismatchNewLocalExitRoot": {
          "STRUCT": [
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "2": {
        "BalanceOverflow": {
          "NEWTYPE": {
            "TYPENAME": "TokenInfo"
          }
        }
      },
      "3": {
        "BalanceUnderflow": {
          "NEWTYPE": {
            "TYPENAME": "TokenInfo"
          }
        }
      },
      "4": {
        "BalanceProofGenerationFailed": {
          "STRUCT": [
            {
              "source": {
                "TYPENAME": "SmtError"
              }
            },
            {
              "token": {
                "TYPENAME": "TokenInfo"
              }
            }
          ]
        }
      },
      "5": {
        "NullifierPathGenerationFailed": {
          "STRUCT": [
            {
              "source": {
                "TYPENAME": "SmtError"
              }
            },
            {
              "global_index": {
                "TYPENAME": "GlobalIndex"
              }
            }
          ]
        }
      },
      "6": {
        "InvalidLocalExitTreeOperation": {
          "NEWTYPE": {
            "TYPENAME": "LocalExitTreeError"
          }
        }
      },
      "7": {
        "L1InfoRootIncorrect": {
          "STRUCT": [
            {
              "leaf_count": "U32"
            },
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "retrieved": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      },
      "8": {
        "InconsistentL1InfoTreeInformation": {
          "STRUCT": [
            {
              "l1_leaf": {
                "OPTION": "U32"
              }
            },
            {
              "l1_info_root": {
                "OPTION": {
                  "TYPENAME": "agglayer_primitives::Digest"
                }
              }
            }
          ]
        }
      },
      "9": {
        "InvalidSmtOperation": {
          "NEWTYPE": {
            "TYPENAME": "SmtError"
          }
        }
      },
      "10": {
        "InconsistentGlobalExitRoot": "UNIT"
      },
      "11": {
        "MissingAggchainVkey": "UNIT"
      },
      "12": {
        "InvalidCustomChainDataLength": {
          "STRUCT": [
            {
              "expected_at_least": "U64"
            },
            {
              "actual": "U64"
            }
          ]
        }
      },
      "13": {
        "MismatchPrevLocalExitRoot": {
          "STRUCT": [
            {
              "computed": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            },
            {
              "declared": {
                "TYPENAME": "agglayer_primitives::Digest"
              }
            }
          ]
        }
      }
    }
  },
  "unified_bridge::Error": {
    "ENUM": {
      "0": {
        "MismatchGlobalIndexInclusionProof": "UNIT"
      },
      "1": {
        "MismatchL1Root": "UNIT"
      },
      "2": {
        "MismatchMER": "UNIT"
      },
      "3": {
        "MismatchRER": "UNIT"
      },
      "4": {
        "InvalidMerklePathLeafToLER": "UNIT"
      },
      "5": {
        "InvalidMerklePathLERToRER": "UNIT"
      },
      "6": {
        "InvalidMerklePathGERToL1Root": "UNIT"
      },
      "7": {
        "InvalidExitNetwork": "UNIT"
      }
    }
  }
}
